### DISCUSSION TITLE
WaitForCompletion and wait for indexes?

### DISCUSSION THREAD

**Discussion by cadilhac** [ravendb/ravendb#19542] [Q&A] [answered]
I am used to use WaitForIndexesAfterSaveChanges after modifying entities on a session so that requerying at the next web request gets the correct result. But I noticed that when using WaitForCompletion after an operation (patching multiple documents in my case), then querying does not give the correct result. Adding WaitForIndexesAfterSaveChanges changed nothing (seems normal since I don't use SaveChanges). How is it possible to wait for the index updates in this case?

Note: I would prefer not to use WaitForNonStaleResults in my query because this is a query at the core of my app and it is done very often.

**Comment by karmeli87** [comment]
Currently there is no built-in option to wait for indexes after patch operation. (you can open a feature request if you like :) )

I can think of a workaround but it is not a nice one (and I didn't actually tried it myself)

You need to pass `QueryOperationOptions` with `RetrieveDetails = true` to the patch operations and use `.WaitForCompletion<BulkOperationResult>` for the result to populate the `Details`.
Those details contain the change vector of each patched document, you need only the last one, since it is the most up to date.
Then, you need to extract the [Etag](https://ravendb.net/docs/article-page/6.2/csharp/server/clustering/replication/change-vector) out of it.

Once you have this, you can do polling on the index with `store.Send(new GetIndexesStatisticsOperation())` and use the `Collections` property to check the `LastProcessedDocumentEtag`.
If your extracted Etag is higher then `LastProcessedDocumentEtag` you know that this index processed all documents of your patch.

**Reply by cadilhac** [reply]
Thank you. Yes, a bit cumbersome. I'll do differently with normal session ops and save.
How can I do an official feature request?

**Reply by karmeli87** [reply]
Here you go: https://issues.hibernatingrhinos.com/issue/RavenDB-23103 :-)

**Reply by cadilhac** [reply]
Feature request in the backlog... I guess a long time will pass until it's implemented, or there is a risk it won't be implemented at all?
I'm surprised this does not seem important. Isn't my use case rather common? I chose to patch because there may be a lot of documents. Then the server redirects to the page that runs a Get query and the results are not correct because the index is not done yet...

**Reply by ppekrol** [reply]
Hi, we are wrapping up 7.0, after the release we will take a peek there.

**Reply by cadilhac** [reply]
I think you implemented in 7.0.3, right? So how to wait for indexes after patching now?

**Reply by ppekrol** [reply, marked as answer]
We added IndexPatchOptions to the QueryOperationOptions. You can use them like this for example:

```csharp
                var operation = store.Operations.Send(new PatchByQueryOperation(new IndexQuery() { Query = "your_patch_query" },
                    new QueryOperationOptions
                    {
                        IndexPatchOptions = new IndexPatchOptions(TimeSpan.FromSeconds(30))
                        {
                            ThrowOnTimeoutInWaitForIndexes = true
                        }
                    }));
```

Following options are available:

```csharp
        /// <summary>
        /// Maximum time to wait for the indexes to become non-stale.
        /// </summary>
        public TimeSpan WaitForIndexesTimeout { get; set; } = DocumentConventions.DefaultWaitForNonStaleResultsTimeout;
        /// <summary>
        /// if true, throw when the timeout is reached, otherwise return normally.
        /// </summary>
        public bool ThrowOnTimeoutInWaitForIndexes { get; set; }
        /// <summary>
        /// A list of index names to wait for. When not set, the server waits only for the indexes it infers from the documents matched by the query or patch. When provided,
        /// the server waits for exactly these index names.
        /// /// </summary>
        public string[] WaitForSpecificIndexes { get; set; }
```

**Reply by cadilhac** [reply]
Thank you. And what is the relation with the existing `operation.WaitForCompletion()` ?
If I wait for indexes, do I still need WaitForCompletion?

**Reply by ppekrol** [reply]
Yes, waiting for completion will conclude after patching finishes and wait for indexes completes.

**Reply by cadilhac** [reply]
So do I understand well by saying that WaitForCompletion is necessary if I really want to wait for the indexes?

**Reply by ppekrol** [reply]
Correct
